File: lib/go/clauses.go
package gencmu

import (
	"fmt"
	"sort"
)

// A rule's clauses against the captures of its alternatives (engine §3.6,
// §9): simplifying a clause for one production, the captures a clause uses
// or mentions, and the checks of a definition as a whole.

// truth is what a condition simplified to: a constant, or open, when it is
// left to be evaluated while parsing.
type truth int8

const (
	open truth = iota
	alwaysTrue
	alwaysFalse
)

// reducedEmpty is the empty set a term reduces to.
var reducedEmpty = &domTerm{Kind: tmEmptySet}

// simplifyCond simplifies a condition for a production that has the
// captures has says it has (engine §3.6): each presence test becomes true
// or false, A ⟹ B becomes B where A is true and true where A is false, a
// true B makes it true and a false one ¬A, and ¬, ∧ and ∨ over a true or
// false part are reduced as logic says. An open result is the condition
// left to evaluate; the one given is returned where nothing changed. A
// reduced part is never evaluated (§10).
func simplifyCond(c *domCond, has func(string) bool) (*domCond, truth) {
	switch c.Kind {
	case cdCaptured:
		if has(c.Rule) {
			return nil, alwaysTrue
		}
		return nil, alwaysFalse
	case cdNot:
		inner, tv := simplifyCond(c.Inner, has)
		switch tv {
		case alwaysTrue:
			return nil, alwaysFalse
		case alwaysFalse:
			return nil, alwaysTrue
		}
		if inner == c.Inner {
			return c, open
		}
		return &domCond{Kind: cdNot, Inner: inner}, open
	case cdAll, cdAny:
		// A false part decides an all, a true part an any; the other
		// constant is dropped.
		decides, drops := alwaysFalse, alwaysTrue
		if c.Kind == cdAny {
			decides, drops = alwaysTrue, alwaysFalse
		}
		var items []*domCond
		changed := false
		for _, it := range c.Items {
			s, tv := simplifyCond(it, has)
			switch tv {
			case decides:
				return nil, decides
			case drops:
				changed = true
				continue
			}
			if s != it {
				changed = true
			}
			items = append(items, s)
		}
		switch {
		case len(items) == 0:
			return nil, drops
		case len(items) == 1:
			return items[0], open
		case !changed:
			return c, open
		}
		return &domCond{Kind: c.Kind, Items: items}, open
	case cdIf:
		premise, tv := simplifyCond(c.Items[0], has)
		if tv == alwaysFalse {
			return nil, alwaysTrue
		}
		then, ttv := simplifyCond(c.Items[1], has)
		switch {
		case tv == alwaysTrue:
			return then, ttv
		case ttv == alwaysTrue:
			return nil, alwaysTrue
		case ttv == alwaysFalse:
			return &domCond{Kind: cdNot, Inner: premise}, open
		case premise == c.Items[0] && then == c.Items[1]:
			return c, open
		}
		return &domCond{Kind: cdIf, Items: []*domCond{premise, then}}, open
	case cdCompare:
		l, r := simplifyTerm(c.Left, has), simplifyTerm(c.Right, has)
		if l == c.Left && r == c.Right {
			return c, open
		}
		return &domCond{Kind: cdCompare, Op: c.Op, Left: l, Right: r}, open
	}
	return c, open
}

// simplifyTerm simplifies a term for a production, as simplifyCond does a
// condition: a guarded term is its term where its condition is true, and
// the empty set where it is false or its term is empty; an empty set is
// dropped from a union, a union of nothing else is empty, and so is an
// intersection with one.
func simplifyTerm(t *domTerm, has func(string) bool) *domTerm {
	if t == nil {
		return nil
	}
	switch t.Kind {
	case tmIf:
		cond, tv := simplifyCond(t.Cond, has)
		switch tv {
		case alwaysFalse:
			return reducedEmpty
		case alwaysTrue:
			return simplifyTerm(t.Items[0], has)
		}
		then := simplifyTerm(t.Items[0], has)
		if then.Kind == tmEmptySet {
			return reducedEmpty
		}
		if cond == t.Cond && then == t.Items[0] {
			return t
		}
		return &domTerm{Kind: tmIf, Cond: cond, Items: []*domTerm{then}}
	case tmUnion, tmIntersection, tmCall:
		// An empty set, written ∅ or left by a guard, is dropped from a
		// union and makes an intersection empty.
		var items []*domTerm
		changed := false
		for _, it := range t.Items {
			s := simplifyTerm(it, has)
			if s != it {
				changed = true
			}
			if s.Kind == tmEmptySet {
				if t.Kind == tmIntersection {
					return reducedEmpty
				}
				if t.Kind == tmUnion {
					changed = true
					continue
				}
			}
			items = append(items, s)
		}
		switch {
		case !changed:
			return t
		case t.Kind == tmUnion && len(items) == 0:
			return reducedEmpty
		case t.Kind == tmUnion && len(items) == 1:
			return items[0]
		}
		return &domTerm{Kind: t.Kind, Str: t.Str, Items: items}
	}
	return t
}

// termCaptures adds the captures a term uses, as values or spans, "" for $;
// a presence test is not a use.
func termCaptures(t *domTerm, into map[string]bool) {
	if t == nil {
		return
	}
	if t.Kind == tmCapture {
		into[t.Str] = true
	}
	if t.Cond != nil {
		condCaptures(t.Cond, into)
	}
	for _, it := range t.Items {
		termCaptures(it, into)
	}
}

func condCaptures(c *domCond, into map[string]bool) {
	switch c.Kind {
	case cdCompare:
		termCaptures(c.Left, into)
		termCaptures(c.Right, into)
	case cdMatches:
		termCaptures(c.Span, into)
	case cdNot:
		condCaptures(c.Inner, into)
	case cdAny, cdAll, cdIf:
		for _, it := range c.Items {
			condCaptures(it, into)
		}
	}
}

// termMentions adds every capture a term mentions, presence tests included.
func termMentions(t *domTerm, into map[string]bool) {
	if t == nil {
		return
	}
	if t.Kind == tmCapture {
		into[t.Str] = true
	}
	if t.Cond != nil {
		condMentions(t.Cond, into)
	}
	for _, it := range t.Items {
		termMentions(it, into)
	}
}

func condMentions(c *domCond, into map[string]bool) {
	switch c.Kind {
	case cdCaptured:
		into[c.Rule] = true
	case cdCompare:
		termMentions(c.Left, into)
		termMentions(c.Right, into)
	case cdMatches:
		termMentions(c.Span, into)
	case cdNot:
		condMentions(c.Inner, into)
	case cdAny, cdAll, cdIf:
		for _, it := range c.Items {
			condMentions(it, into)
		}
	}
}

// altCaptures maps each capture of an alternative to its position among the
// items of its top level; $ is at -1.
func altCaptures(a *domAlt) map[string]int {
	out := map[string]int{"": -1}
	items := []*domExpr{a.Expr}
	if a.Expr.Kind == exSeq {
		items = a.Expr.Items
	}
	for i, it := range items {
		if it.Kind == exCapture {
			out[it.Name] = i
		}
	}
	return out
}

// usesAll says whether a clause, simplified, uses only captures has has.
func usesAll(names map[string]bool, has func(string) bool) (string, bool) {
	for n := range names {
		if !has(n) {
			return n, false
		}
	}
	return "", true
}

// definitionProblem is why a definition, a rule's alternatives with the
// clauses written with them, cannot be read (engine §9), or "". The DOM's
// shape must already be sound.
func definitionProblem(r *domRule) string {
	alts := make([]map[string]int, len(r.Alternatives))
	for i, a := range r.Alternatives {
		alts[i] = altCaptures(a)
	}
	var items []*domEmitItem
	if r.Emit != nil {
		items = r.Emit.Items
	}
	// A capture no alternative captures, wherever it is mentioned.
	mentioned := map[string]bool{}
	termMentions(r.Tags, mentioned)
	for _, c := range r.Conditions {
		condMentions(c, mentioned)
	}
	for _, a := range r.Alternatives {
		termMentions(a.Tags, mentioned)
	}
	for _, it := range items {
		if !it.IsInsert {
			mentioned[it.Capture] = true
		}
		termMentions(it.Tags, mentioned)
	}
	for _, name := range sortedKeys(mentioned) {
		found := false
		for _, caps := range alts {
			if _, ok := caps[name]; ok {
				found = true
				break
			}
		}
		if !found {
			return fmt.Sprintf("$%s is captured by no alternative of %s", name, r.Name)
		}
	}
	// A condition that applies to no alternative.
	for _, c := range r.Conditions {
		applies := false
		for _, caps := range alts {
			has := hasIn(caps)
			s, tv := simplifyCond(c, has)
			if tv == alwaysTrue {
				continue
			}
			if tv == alwaysFalse {
				applies = true
				break
			}
			used := map[string]bool{}
			condCaptures(s, used)
			if _, ok := usesAll(used, has); ok {
				applies = true
				break
			}
		}
		if !applies {
			return fmt.Sprintf("a condition of %s applies to no alternative", r.Name)
		}
	}
	unguarded := func(t *domTerm, has func(string) bool) string {
		if t == nil {
			return ""
		}
		used := map[string]bool{}
		termCaptures(simplifyTerm(t, has), used)
		if name, ok := usesAll(used, has); !ok {
			return fmt.Sprintf("a tag term of %s uses $%s, which an alternative lacks; guard it with $%s ⟹", r.Name, name, name)
		}
		return ""
	}
	for i, a := range r.Alternatives {
		caps := alts[i]
		has := hasIn(caps)
		// The tags an alternative's constituent carries serve it.
		for _, t := range []*domTerm{r.Tags, a.Tags} {
			if msg := unguarded(t, has); msg != "" {
				return msg
			}
		}
		if r.Emit == nil {
			continue
		}
		// What is left of the emission for this alternative: something, in
		// the order its captures stand, each item's tags using only what it
		// has.
		last, left := -2, 0
		for _, it := range items {
			if !it.IsInsert && !has(it.Capture) {
				continue
			}
			left++
			if it.IsInsert {
				continue
			}
			if it.Capture != "" {
				at := caps[it.Capture]
				if at < last {
					return fmt.Sprintf("%%emits of %s lists captures out of the order they stand in", r.Name)
				}
				last = at
			}
			if msg := unguarded(it.Tags, has); msg != "" {
				return msg
			}
		}
		// Only a rule that lists items can leave nothing; ε lists none.
		if left == 0 && len(items) > 0 {
			return fmt.Sprintf("%%emits of %s leaves an alternative nothing to emit; a rule that emits nothing says %%emits ε", r.Name)
		}
	}
	// An inserted tag's anchor, the capture listed next after it, is one
	// every alternative has.
	for i, it := range items {
		if !it.IsInsert {
			continue
		}
		for _, next := range items[i+1:] {
			if next.IsInsert {
				continue
			}
			for _, caps := range alts {
				if _, ok := caps[next.Capture]; !ok {
					return fmt.Sprintf("%%emits of %s inserts a tag before $%s, which an alternative lacks", r.Name, next.Capture)
				}
			}
			break
		}
	}
	return ""
}

func sortedKeys(m map[string]bool) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

func hasIn(caps map[string]int) func(string) bool {
	return func(name string) bool {
		_, ok := caps[name]
		return ok
	}
}

File: lib/rust/src/clauses.rs
//! A rule's clauses against the captures of its alternatives (engine §3.6,
//! §9): simplifying a clause for a production, the captures a clause uses,
//! and the checks of a definition as a whole.

use crate::dom::{Alternative, Arg, Cond, EmitItem, Expr, RuleDef, Term};

/// A condition simplified for a production: decided, or still to evaluate.
#[derive(Debug, Clone, PartialEq)]
pub(crate) enum Simple {
    True,
    False,
    Cond(Cond),
}

/// Simplifies a condition for a production that has the captures `has`
/// (engine §3.6): each presence test becomes true or false, and `⟹`, `¬`,
/// `∧` and `∨` over a true or false part are reduced.
pub(crate) fn simplify_cond(cond: &Cond, has: &dyn Fn(&str) -> bool) -> Simple {
    match cond {
        Cond::Captured(name) => {
            if has(name) {
                Simple::True
            } else {
                Simple::False
            }
        }
        Cond::Not(inner) => match simplify_cond(inner, has) {
            Simple::True => Simple::False,
            Simple::False => Simple::True,
            Simple::Cond(inner) => Simple::Cond(Cond::Not(Box::new(inner))),
        },
        Cond::All(items) => {
            let mut left = Vec::new();
            for item in items {
                match simplify_cond(item, has) {
                    Simple::False => return Simple::False,
                    Simple::True => {}
                    Simple::Cond(item) => left.push(item),
                }
            }
            match left.len() {
                0 => Simple::True,
                1 => Simple::Cond(left.pop().expect("one condition")),
                _ => Simple::Cond(Cond::All(left)),
            }
        }
        Cond::Any(items) => {
            let mut left = Vec::new();
            for item in items {
                match simplify_cond(item, has) {
                    Simple::True => return Simple::True,
                    Simple::False => {}
                    Simple::Cond(item) => left.push(item),
                }
            }
            match left.len() {
                0 => Simple::False,
                1 => Simple::Cond(left.pop().expect("one condition")),
                _ => Simple::Cond(Cond::Any(left)),
            }
        }
        Cond::If(antecedent, consequent) => {
            let antecedent = match simplify_cond(antecedent, has) {
                Simple::False => return Simple::True,
                Simple::True => return simplify_cond(consequent, has),
                Simple::Cond(antecedent) => antecedent,
            };
            match simplify_cond(consequent, has) {
                Simple::True => Simple::True,
                Simple::False => Simple::Cond(Cond::Not(Box::new(antecedent))),
                Simple::Cond(consequent) => Simple::Cond(Cond::If(Box::new(antecedent), Box::new(consequent))),
            }
        }
        Cond::Compare(op, left, right) => {
            Simple::Cond(Cond::Compare(op.clone(), simplify_value(left, has), simplify_value(right, has)))
        }
        Cond::Matches(..) | Cond::Initial(_) => Simple::Cond(cond.clone()),
    }
}

/// Simplifies a term for a production (engine §3.6): `None` where it is
/// the empty set, written `∅` or left by a guard that does not hold. The
/// empty set is dropped from a union, makes an intersection empty, and
/// makes a guarded term empty.
pub(crate) fn simplify_term(term: &Term, has: &dyn Fn(&str) -> bool) -> Option<Term> {
    match term {
        Term::EmptySet => None,
        Term::If(cond, then) => match simplify_cond(cond, has) {
            Simple::False => None,
            Simple::True => simplify_term(then, has),
            Simple::Cond(cond) => Some(Term::If(Box::new(cond), Box::new(simplify_term(then, has)?))),
        },
        Term::Union(items) => {
            let mut left: Vec<Term> = items.iter().filter_map(|item| simplify_term(item, has)).collect();
            match left.len() {
                0 => None,
                1 => left.pop(),
                _ => Some(Term::Union(left)),
            }
        }
        Term::Intersection(items) => {
            let mut left = Vec::with_capacity(items.len());
            for item in items {
                left.push(simplify_term(item, has)?);
            }
            Some(Term::Intersection(left))
        }
        _ => Some(term.clone()),
    }
}

/// A term simplified for a production, the empty set written out.
pub(crate) fn simplify_value(term: &Term, has: &dyn Fn(&str) -> bool) -> Term {
    simplify_term(term, has).unwrap_or(Term::EmptySet)
}

fn term_captures<'a>(term: &'a Term, out: &mut Vec<&'a str>) {
    match term {
        Term::Capture(name) => out.push(name),
        Term::Union(items) | Term::Intersection(items) => items.iter().for_each(|item| term_captures(item, out)),
        Term::Call(_, args) => {
            for arg in args {
                if let Arg::Term(term) = arg {
                    term_captures(term, out);
                }
            }
        }
        Term::If(cond, then) => {
            cond_captures(cond, out);
            term_captures(then, out);
        }
        Term::Literal(_) | Term::Weak(_) | Term::EmptySet => {}
    }
}

fn cond_captures<'a>(cond: &'a Cond, out: &mut Vec<&'a str>) {
    match cond {
        Cond::Compare(_, left, right) => {
            term_captures(left, out);
            term_captures(right, out);
        }
        Cond::Matches(span, _) | Cond::Initial(span) => term_captures(span, out),
        Cond::Not(inner) => cond_captures(inner, out),
        Cond::Any(items) | Cond::All(items) => items.iter().for_each(|item| cond_captures(item, out)),
        Cond::If(antecedent, consequent) => {
            cond_captures(antecedent, out);
            cond_captures(consequent, out);
        }
        Cond::Captured(_) => {}
    }
}

/// The captures a term uses as values or spans, presence tests aside.
pub(crate) fn term_uses(term: &Term) -> Vec<&str> {
    let mut out = Vec::new();
    term_captures(term, &mut out);
    out
}

/// The captures a condition uses as values or spans, presence tests aside.
pub(crate) fn cond_uses(cond: &Cond) -> Vec<&str> {
    let mut out = Vec::new();
    cond_captures(cond, &mut out);
    out
}

fn term_presences<'a>(term: &'a Term, out: &mut Vec<&'a str>) {
    match term {
        Term::Union(items) | Term::Intersection(items) => items.iter().for_each(|item| term_presences(item, out)),
        Term::If(cond, then) => {
            cond_presences(cond, out);
            term_presences(then, out);
        }
        _ => {}
    }
}

fn cond_presences<'a>(cond: &'a Cond, out: &mut Vec<&'a str>) {
    match cond {
        Cond::Captured(name) => out.push(name),
        Cond::Compare(_, left, right) => {
            term_presences(left, out);
            term_presences(right, out);
        }
        Cond::Not(inner) => cond_presences(inner, out),
        Cond::Any(items) | Cond::All(items) => items.iter().for_each(|item| cond_presences(item, out)),
        Cond::If(antecedent, consequent) => {
            cond_presences(antecedent, out);
            cond_presences(consequent, out);
        }
        Cond::Matches(..) | Cond::Initial(_) => {}
    }
}

/// The captures of an alternative's top level, each with its position;
/// `$` is not among them.
pub(crate) fn alternative_captures(alternative: &Alternative) -> Vec<(&str, usize)> {
    let top: &[Expr] = match &alternative.expr {
        Expr::Seq(items) => items,
        other => std::slice::from_ref(other),
    };
    top.iter()
        .enumerate()
        .filter_map(|(position, item)| match item {
            Expr::Capture(name, _) => Some((name.as_str(), position)),
            _ => None,
        })
        .collect()
}

/// Why a definition, a rule's alternatives with its own clauses, is an
/// error of the document (engine §9), or `None`.
pub(crate) fn definition_problem(rule: &RuleDef) -> Option<String> {
    let alternatives: Vec<Vec<(&str, usize)>> = rule.alternatives.iter().map(alternative_captures).collect();
    let captures_of = |index: usize| {
        let captures = &alternatives[index];
        move |name: &str| name.is_empty() || captures.iter().any(|(captured, _)| *captured == name)
    };
    let any_has = |name: &str| name.is_empty() || alternatives.iter().flatten().any(|(captured, _)| *captured == name);
    let items: &[EmitItem] = rule.emit.as_deref().unwrap_or(&[]);
    // A constituent that does not count cannot sound like its text (§9).
    if rule.verbatim && rule.emit.is_some() && items.is_empty() {
        return Some(format!("{} is verbatim and emits ε", rule.name));
    }

    // A capture mentioned anywhere, presence tests included, that no
    // alternative captures.
    let mut mentioned: Vec<&str> = Vec::new();
    for term in rule.tags.iter().chain(rule.alternatives.iter().filter_map(|alternative| alternative.tags.as_ref())) {
        term_captures(term, &mut mentioned);
        term_presences(term, &mut mentioned);
    }
    for cond in &rule.conditions {
        cond_captures(cond, &mut mentioned);
        cond_presences(cond, &mut mentioned);
    }
    for item in items {
        match item {
            EmitItem::Capture(name, tags) => {
                mentioned.push(name);
                if let Some(term) = tags {
                    term_captures(term, &mut mentioned);
                    term_presences(term, &mut mentioned);
                }
            }
            EmitItem::Insert(_) => {}
        }
    }
    if let Some(name) = mentioned.iter().find(|name| !any_has(name)) {
        return Some(format!("${name} is captured by no alternative of {}", rule.name));
    }

    // A condition that applies to no alternative.
    for cond in &rule.conditions {
        let applies = (0..alternatives.len()).any(|index| {
            let has = captures_of(index);
            match simplify_cond(cond, &has) {
                Simple::True => false,
                Simple::False => true,
                Simple::Cond(simple) => cond_uses(&simple).iter().all(|name| has(name)),
            }
        });
        if !applies {
            return Some(format!("a condition of {} applies to no alternative", rule.name));
        }
    }

    let unguarded = |name: &str| {
        format!("a tag term of {} uses ${name}, which an alternative lacks; guard it with ${name} ⟹", rule.name)
    };
    for (index, alternative) in rule.alternatives.iter().enumerate() {
        let has = captures_of(index);
        for term in [rule.tags.as_ref(), alternative.tags.as_ref()].into_iter().flatten() {
            if let Some(simple) = simplify_term(term, &has) {
                if let Some(missing) = term_uses(&simple).into_iter().find(|name| !has(name)) {
                    return Some(unguarded(missing));
                }
            }
        }
        if rule.emit.is_none() {
            continue;
        }
        let present: Vec<&EmitItem> = items
            .iter()
            .filter(|item| match item {
                EmitItem::Capture(name, _) => has(name),
                EmitItem::Insert(_) => true,
            })
            .collect();
        // Nothing left to emit is an error only where the rule lists items:
        // `%emits ε` lists none (§9).
        if present.is_empty() && !items.is_empty() {
            return Some(format!("%emits of {} leaves an alternative nothing to emit", rule.name));
        }
        let positions: Vec<usize> = present
            .iter()
            .filter_map(|item| match item {
                EmitItem::Capture(name, _) if !name.is_empty() => {
                    alternatives[index].iter().find(|(captured, _)| captured == name).map(|(_, position)| *position)
                }
                _ => None,
            })
            .collect();
        if positions.windows(2).any(|pair| pair[1] < pair[0]) {
            return Some(format!("%emits of {} lists captures out of the order they stand in", rule.name));
        }
        for item in &present {
            if let EmitItem::Capture(_, Some(term)) = item {
                if let Some(simple) = simplify_term(term, &has) {
                    if let Some(missing) = term_uses(&simple).into_iter().find(|name| !has(name)) {
                        return Some(unguarded(missing));
                    }
                }
            }
        }
    }

    // An inserted tag whose anchor, the capture listed next after it, some
    // alternative lacks.
    for (index, item) in items.iter().enumerate() {
        if !matches!(item, EmitItem::Insert(_)) {
            continue;
        }
        let anchor = items[index + 1..].iter().find_map(|item| match item {
            EmitItem::Capture(name, _) => Some(name.as_str()),
            EmitItem::Insert(_) => None,
        });
        if let Some(anchor) = anchor {
            if !(0..alternatives.len()).all(|alternative| captures_of(alternative)(anchor)) {
                return Some(format!(
                    "%emits of {} inserts a tag before ${anchor}, which an alternative lacks",
                    rule.name
                ));
            }
        }
    }
    None
}
